providers/websockets/gate/ws_data_handler.go

```diff
@@ -0,0 +1,127 @@
+		return nil, fmt.Errorf("invalid provider config %s", err)
+	}
+
+	if !cfg.WebSocket.Enabled {
+		return nil, fmt.Errorf("web socket is not enabled for provider %s", cfg.Name)
+	}
+
+	if cfg.Name != Name {
+		return nil, fmt.Errorf("invalid provider name %s", cfg.Name)
+	}
+
+	return &WebsocketDataHandler{
+		cfg:    cfg,
+		logger: logger.With(zap.String("web_socket_data_handler", Name)),
+	}, nil
+}
+
+// HandleMessage is used to handle a message received from the data provider. The Gate.io
+// provider sends two types of messages:
+//
+//  1. Subscribe response message. The subscribe response message is used to determine if
+//     the subscription was successful.
+//  2. Ticker stream message. This is sent when a ticker update is received from the
+//     Gate.io web socket API.
+func (h *WebsocketDataHandler) HandleMessage(
+	message []byte,
+) (providertypes.GetResponse[oracletypes.CurrencyPair, *big.Int], []handlers.WebsocketEncodedMessage, error) {
+	var (
+		resp         providertypes.GetResponse[oracletypes.CurrencyPair, *big.Int]
+		baseMessage  BaseMessage
+		subResponse  SubscribeResponse
+		tickerStream TickerStream
+	)
+
+	// Attempt to unmarshal the message into a base message. This is used to determine the type
+	// of message that was received.
+	if err := json.Unmarshal(message, &baseMessage); err != nil {
+		h.logger.Debug("unable to unmarshal message into base message", zap.Error(err))
+		return resp, nil, err
+	}
+
+	switch Event(baseMessage.Event) {
+	case EventSubscribe:
+		if err := json.Unmarshal(message, &subResponse); err != nil {
+			h.logger.Debug("unable to unmarshal message into subscribe response", zap.Error(err))
+			return resp, nil, err
+		}
+
+		// handle subscription
+		updateMsg, err := h.parseSubscribeResponse(subResponse)
+		return resp, updateMsg, err
+
+	case EventUpdate:
+		if err := json.Unmarshal(message, &tickerStream); err != nil {
+			h.logger.Debug("unable to unmarshal message into ticker stream", zap.Error(err))
+			return resp, nil, err
+		}
+
+		// update pair info
+		resp, err := h.parseTickerStream(tickerStream)
+		return resp, nil, err
+
+	default:
+		h.logger.Debug("received unknown message", zap.String("message", string(message)))
+		return resp, nil, fmt.Errorf("unknown message type %s", baseMessage.Event)
+	}
+}
+
+// CreateMessages is used to create an initial subscription message to send to the data provider.
+// Only the currency pairs that are specified in the config are subscribed to. The only channel
+// that is subscribed to is the tickers channel - which supports spot markets.
+func (h *WebsocketDataHandler) CreateMessages(
+	cps []oracletypes.CurrencyPair,
+) ([]handlers.WebsocketEncodedMessage, error) {
+	symbols := make([]string, 0)
+
+	for _, cp := range cps {
+		market, ok := h.cfg.Market.CurrencyPairToMarketConfigs[cp.String()]
+		if !ok {
+			h.logger.Debug("market not found for currency pair", zap.String("currency_pair", cp.String()))
+			continue
+		}
```

Comment:
nit: I think we should error if the market is unsupported, this means something in the config is not correct. We should fix the other providers to account for this.